### PR TITLE
Metrics: Added metrics configuration

### PR DESCRIPTION
**Pull request by eloycoto**
This commits adds the new metrics information on the spec. The value
will be `port` and `path`.

List of changes:

- Changed CRD definition.
- Update workload API response definition.
- If port is not defined, no metrics will be used.

Related to: ECOPROJECT-425

Signed-off-by: Eloy Coto <eloy.coto@acalustra.com>

### CODE REVIEW

File: api/v1alpha1/edgedeployment_types.go [hunk 1]
```diff
@@ -32,12 +32,18 @@ type EdgeDeploymentSpec struct {
...
 	Data            *DataConfiguration            `json:"data,omitempty"`
 	ImageRegistries *ImageRegistriesConfiguration `json:"imageRegistries,omitempty"`
+	Metrics         *PrometheusMetrics            `json:"metrics,omitempty"`
 }
 
 type ImageRegistriesConfiguration struct {
 	AuthFileSecret *ObjectRef `json:"secretRef,omitempty"`
 }
 
+type PrometheusMetrics struct {
```
Comment (jakub-dzon) [inline api/v1alpha1/edgedeployment_types.go:42]: Let's keep it technology-agnostic: `MetricsConfiguration`

File: internal/yggdrasil/yggdrasil.go [hunk 2]
```diff
@@ -405,6 +405,13 @@ func (h *Handler) toWorkloadList(ctx context.Context, logger logr.Logger, deploy
 				AuthFile: authFile,
 			}
 		}
+
+		if spec.Metrics != nil && spec.Metrics.Port > 0 {
+			workload.Metrics = &models.Metrics{
+				Path: spec.Metrics.Path,
```
Comment (jakub-dzon) [inline internal/yggdrasil/yggdrasil.go:411]: What if `Port>0` and `Path==""`?

**Review by jakub-dzon — CHANGES_REQUESTED** [review]

File: internal/yggdrasil/yggdrasil.go (hunk 2 above)
Comment (eloycoto) [reply internal/yggdrasil/yggdrasil.go:411]: I guess that path empty is /, no?

**Comment by eloycoto** [comment]
@jakub-dzon  and all, what about if one Deployment defines two pods? will require the two containers to listen at the same port/path?

Regards

File: internal/yggdrasil/yggdrasil.go (hunk 2 above)
Comment (jakub-dzon) [reply internal/yggdrasil/yggdrasil.go:411]: :+1:

**Comment by jakub-dzon** [comment]
Good point. We don't support deployments and might not be implementing that support at all, but we might have similar problem already - pod can have multiple containers. What if there is a need to scrape metrics from more than one? What if we need to scrape from only one of many? IMO we can enforce same path/port for all as the simplest solution, but selective metric scraping might be a problem.

**Comment by jakub-dzon** [comment]
We might also need to allow users to define scraping interval to allow for different metrics resolutions.

**Comment by eloycoto** [comment]
Added the changes + Kubebuilder annotations for min/max.

File: internal/1.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+11
```
Comment (jakub-dzon) [inline internal/1.txt:1]: That file has probably been added by mistake.

File: internal/images/gomock_reflect_782103550/prog.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
+package main
```
Comment (jakub-dzon) [inline internal/images/gomock_reflect_782103550/prog.go:1]: That file has probably been added by mistake.

**Review by jakub-dzon — CHANGES_REQUESTED** [review]
There are some stray temporary files that need to be removed.

**Review by jakub-dzon — APPROVED** [review]
LGTM. 
Please rebase.

**Review by pkliczewski — APPROVED** [review]

File: swagger.yaml [hunk 5]
```diff
@@ -267,6 +270,36 @@ definitions:
...
         type: string
 
+  metrics:
+    type: object
+    properties:
+      path:
+        description: Path to use when retrieving metrics
+        type: string
+      port:
+        type: integer
```
Comment (masayag) [inline swagger.yaml:280]: WDYT of including also `format: int32` which is suitable for the port range?

File: api/v1alpha1/edgedeployment_types.go [hunk 6]
```diff
@@ -26,18 +26,50 @@ import (
...
 
+type MetricsConfigEntity struct {
+	// Path to use when retrieving metrics
+	// +kubebuilder:default=/
+	Path string `json:"path,omitempty"`
+
+	// Port to use when retrieve the metrics
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=65535
+	Port int `json:"port,omitempty"`
```
Comment (masayag) [inline api/v1alpha1/edgedeployment_types.go:50]: int32 should fit this purpose.

File: api/v1alpha1/edgedeployment_types.go [hunk 7]
```diff
@@ -26,18 +26,50 @@ import (
...
+
+type ContainerMetricsConfiguration struct {
+	// Path to use when retrieving metrics
+	// +kubebuilder:default=/
+	Path string `json:"path,omitempty"`
+
+	// Port to use when retrieve the metrics
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=65535
+	Port int `json:"port,omitempty"`
```
Comment (masayag) [inline api/v1alpha1/edgedeployment_types.go:63]: int32 should fit this purpose.

File: internal/yggdrasil/yggdrasil.go [hunk 8]
```diff
@@ -422,6 +422,26 @@ func (h *Handler) toWorkloadList(ctx context.Context, logger logr.Logger, deploy
...
+			for container, metricConf := range spec.Metrics.Containers {
+				containers[container] = models.ContainerMetrics{
+					Disabled: metricConf.Disabled,
+					Port:     int64(metricConf.Port),
+					Path:     metricConf.Path,
+				}
+				addedContainers = true
+			}
+			if addedContainers {
+				workload.Metrics.Containers = containers
```
Comment (masayag) [inline internal/yggdrasil/yggdrasil.go:442]: on the worker side workload.Metrics.Containers with `nil` or zero value behaves differently?

Comment (eloycoto) [reply internal/yggdrasil/yggdrasil.go:442]: no, will range over Metrics.Containers.

File: swagger.yaml (hunk 5 above)
Comment (eloycoto) [reply swagger.yaml:280]: Done! Sorry!

**Review by masayag — APPROVED** [review]

**Pull request merged**